src/main/java/net/openhft/chronicle/bytes/internal/BytesInternal.java

```diff
@@ -319,37 +319,75 @@ public static boolean startsWith(@NotNull BytesStore a, @NotNull BytesStore b)
         throwExceptionIfReleased(a);
         throwExceptionIfReleased(b);
         final long bRealReadRemaining = b.realReadRemaining();
-
         if (a.realReadRemaining() < bRealReadRemaining)
             return false;
+
         try {
-            long aPos = a.readPosition();
-            long bPos = b.readPosition();
-            long length = bRealReadRemaining;
-            long i;
-            for (i = 0; i < length - 7; i += 8) {
-                if (a.readLong(aPos + i) != b.readLong(bPos + i))
-                    return false;
-            }
-            if (i < length - 3) {
-                if (a.readInt(aPos + i) != b.readInt(bPos + i))
-                    return false;
-                i += 4;
-            }
-            if (i < length - 1) {
-                if (a.readShort(aPos + i) != b.readShort(bPos + i))
-                    return false;
-                i += 2;
-            }
-            if (i < length) {
-                return a.readByte(aPos + i) == b.readByte(bPos + i);
+            if (a instanceof HasUncheckedRandomDataInput && b instanceof HasUncheckedRandomDataInput) {
```

Comment:
We could add a check for a certain length threshold here before we do the intstanceof.

Reply:
Would it be faster if the interface had a startsWith method, then you'd only have to `instanceof` the RHS? (The LHS would just use whatever implementation it implemented)

E.g. a.startsWith(b)

Ignore me if this is gibberish.

Reply:
The best would be if we had "traits" and then we could just select two different views: one checked and one unchecked. I.e. we separate the concerns of memory and access.

The solution above only caters to the case both can be optimized when, in reality, there are four combinations. The other three will use checked access.

Related: https://stackoverflow.com/questions/103564/the-performance-impact-of-using-instanceof-in-java

Reply:
But your suggestion would work and we could override the relevant classes. It would indeed mean one less `instanceof `. There is already an implementation in each concrete class (inherited from `BytesStore`)

Reply:
You can just add a method overload, i.e. public static boolean startsWith(@NotNull HasUncheckedRandomDataInput a, @NotNull HasUncheckedRandomDataInput b) - and then optionally call it here.